feat: add `ToPrint` to framework for better MLIR printing

luisacicolini opened the pull request:
The current `toString` implementation is not sufficiently generic to print correct MLIR syntax. This PR introduces a new `toPrint` infrastructure that enables accurate printing of MLIR syntax.

We should investigate if and how much of the `toString` infrastructure is still needed and eliminate this redundancy.

This PR imports the changes to `Framework.lean`, to `Base.lean`, and to `Opt.lean` from #1538. 

Authored by @salinhkuhn.

luisacicolini committed:
chore: import changes to framework

luisacicolini committed:
chore: minimal changes

luisacicolini committed:
chore: minimal changes

luisacicolini committed:
chore: minimize diff

tobiasgrosser commented:
@luisacicolini, you probably want to give this PR a title and body that describes what it actually does, e.g, add printing support ...

luisacicolini committed:
chore: all printing in the same section

luisacicolini committed:
chore: minimize diff

luisacicolini committed:
chore: printing infra in base file

luisacicolini committed:
chore: update opt

salinhkuhn commented on SSA/Projects/InstCombine/LLVM/Opt.lean:
```diff
@@ -51,5 +51,5 @@ def mainCmd := `[Cli|
     ]
 
-def main (args : List String): IO UInt32 :=
+unsafe def main (args : List String): IO UInt32 :=
```
not anymore, I made CSE safe, so its not needed anymore. Locally on my branch I can remove it and it builds.

luisacicolini committed:
chore: remove unsafe

luisacicolini committed:
chore: address comments

luisacicolini commented on SSA/Core/Framework.lean:
```diff
@@ -303,6 +329,48 @@ instance : ToString (Expr d Γ eff t) where toString := Expr.toString
+    s!"  %{_root_.repr <|(Γ.length)} = {Expr.toPrint e }" ++ "\n" ++
+    Com.ToPrintBody body
+
+/- `Com.toString` implements a toString instance for the type `Com`.  -/
+partial def Com.toPrint (com : Com d Γ eff t) : String :=
+   "builtin.module { \n"
+  ++ DialectPrint.printFunc t ++ ((formatFormalArgListTuplePrint Γ.toList)) ++ ":" ++ "\n"
+  ++ (Com.ToPrintBody com) ++
+   "\n }"
```
resolving the comment as we spoke about this in the morning.

luisacicolini committed:
chore: address comments

tobiasgrosser commented:
> I am especially unsure about the changes to Opt, as making unsafe def main does not seem encouraging.

The changes to `opt` are not in this PR anymore. Hence, the commit message should be updated.

tobiasgrosser commented on SSA/Core/Framework.lean:
```diff
@@ -303,6 +319,77 @@ instance : ToString (Expr d Γ eff t) where toString := Expr.toString
 
 end ToString
 
+/- # ToString instances for Com and Expr  -/
+section DialectPrint
```
This comment seems incorrect. These are not the `toString` instances, no?

salinhkuhn replied:
yes

tobiasgrosser replied:
Can someone update the comment and fix this?

luisacicolini replied:
done

tobiasgrosser commented on SSA/Core/Framework.lean:
```diff
@@ -303,6 +319,77 @@ instance : ToString (Expr d Γ eff t) where toString := Expr.toString
+
+/--
+Converts an expression within a dialect to its MLIR string representation.
+Since MLIR generic syntax uses "" to denote operations, this function, this functions uses
```
```suggestion
Since MLIR generic syntax uses "" to denote operations, this functions uses
```

Also, what does `""` refer to?

luisacicolini committed:
chore: accept docstrings suggestion

Co-authored-by: Tobias Grosser <tobias@grosser.es>

salinhkuhn commented:
> I don't like that this PR adds a _third_ distinct way of printing stuff, can the new printing infra replace the existing `ToString` instance?

salinhkuhn commented:
> I don't like that this PR adds a _third_ distinct way of printing stuff, can the new printing infra replace the existing `ToString` instance?

it wanted to be more customisable than toString. I assume that there is a certain amount of standard of what a toString instance in Lean should print ? I aimed for toPrint to not have any restrictions on how the dialect object should be printed

tobiasgrosser commented on SSA/Core/Framework.lean:
```diff
@@ -303,6 +319,77 @@ instance : ToString (Expr d Γ eff t) where toString := Expr.toString
+    Com.ToPrintBody body
+
+/--
+  `Com.toString` implements a `toString` instance for the type `Com`.
```
What does `yes, DialectPrint` mean? Does it mean the above is correct or it should be changed to sth else. If it should be changed, can you propose another sentence?

luisacicolini replied:
I updated the docstring, AFAIU this should be correct now. @salinhkuhn please confirm whether this should continue to be a `partial def` or we should instead instantiate `DialectPrint`.

tobiasgrosser committed:
Merge remote-tracking branch 'origin/main' into llvm-riscv-merging-1

tobiasgrosser committed:
Merge remote-tracking branch 'origin/main' into llvm-riscv-merging-1

luisacicolini committed:
chore: update comment

luisacicolini committed:
chore: comment

luisacicolini committed:
chore: clarify docstring

tobiasgrosser commented on SSA/Core/Framework.lean:
```diff
@@ -302,6 +318,77 @@ instance : ToString (Expr d Γ eff t) where toString := Expr.toString
+  This function recursively converts the body of a `Com` into its string representation.
+  Each bound variable is printed with its index and corresponding expression.
+-/
+partial def Com.ToPrintBody : Com d Γ eff t → String
+  | .ret v => s!"  \"{DialectPrint.printReturn t}\"({_root_.repr v }) : ({DialectPrint.printTy t}) -> ()"
+  | .var e body =>
+    s!"  %{_root_.repr <|(Γ.length)} = {Expr.toPrint e }" ++ "\n" ++
+    Com.ToPrintBody body
+
+/--
+  `Com.toPrint` implements a `toPrint` instance for the type `Com`.
+  This has a more general behaviour than `toString` and allows customizing the
+  printing of dialect objects.
+-/
+partial def Com.toPrint (com : Com d Γ eff t) : String :=
+   "builtin.module { \n"
+  ++ DialectPrint.printFunc t ++ ((formatFormalArgListTuplePrint Γ.toList)) ++ ":" ++ "\n"
+  ++ (Com.ToPrintBody com) ++
```
```suggestion
partial def Com.toPrintBody : Com d Γ eff t → String
  | .ret v => s!"  \"{DialectPrint.printReturn t}\"({_root_.repr v }) : ({DialectPrint.printTy t}) -> ()"
  | .var e body =>
    s!"  %{_root_.repr <|(Γ.length)} = {Expr.toPrint e }" ++ "\n" ++
    Com.ToPrintBody body

/--
  `Com.toPrint` implements a `toPrint` instance for the type `Com`.
  This has a more general behaviour than `toString` and allows customizing the
  printing of dialect objects.
-/
partial def Com.toPrint (com : Com d Γ eff t) : String :=
   "builtin.module { \n"
  ++ DialectPrint.printFunc t ++ ((formatFormalArgListTuplePrint Γ.toList)) ++ ":" ++ "\n"
  ++ (Com.toPrintBody com) ++
```

It seems we typically write `toPrint`, so let's use this consistently.

tobiasgrosser commented on SSA/Projects/InstCombine/Base.lean:
```diff
@@ -393,6 +409,14 @@ instance : ToString (MTy φ) where
 
+instance : DialectPrint (LLVM) where
+  printOpName
+  | op => "llvm."++toString op
```
```suggestion
  | op => "llvm." ++ toString op
```

We typically use whitespace around `++`.

tobiasgrosser commented on SSA/Core/Framework.lean:
```diff
@@ -302,6 +318,77 @@ instance : ToString (Expr d Γ eff t) where toString := Expr.toString
 
 end ToString
 
+/- # ToPrint instances for Com and Expr  -/
```
``` suggestion
/- # `toPrint` instances for Com and Expr  -/
```

For consistency, we could use the same casing across the project.

tobiasgrosser commented on SSA/Core/Framework.lean:
```diff
@@ -140,6 +140,22 @@ Another option, which would be more complicated, but correct by construction, is
+/--
+DialectPrint includes the functions to print the components of a dialect.
+-/
+class DialectPrint (d : Dialect) where
```
There is a naming mismatch here. We talk later about the `ToPrint` instances, but call the class `DialectPrint`. Should we just name it `ToPrint`?

tobiasgrosser replied:
I renamed this for now to `ToPrint`.

tobiasgrosser reviewed (approved):
I am overall happy with this PR. Let's address the minor glitches and then get this shipped. We can independently think about if/how to remove the `ToString` instances, I guess.

tobiasgrosser committed:
Merge remote-tracking branch 'origin/main' into llvm-riscv-merging-1

tobiasgrosser committed:
Only run evaluation on 'main' (#1551)

Our evaluation is currently to slow to run on all pushes. Let's run it
on master only to make sure the red crosses do not add unnecessary
noise.

tobiasgrosser committed:
evaluation: try to run on GitHub runners to see if this is fast enough (#1552)

tobiasgrosser committed:
evaluation: try to run on GitHub runners to see if this is fast enough (#1553)

tobiasgrosser committed:
Update SSA/Projects/InstCombine/Base.lean

tobiasgrosser committed:
Update SSA/Projects/InstCombine/Base.lean

tobiasgrosser committed:
Update SSA/Core/Framework.lean

tobiasgrosser committed:
eval: continue on error for test-case compare (#1554)

Eval is currently failing. Let's disable it to at least have a green CI
for now.

tobiasgrosser committed:
chore: replace eval with #guard_msgs (#1555)

This reduces info noise and ensures the output is actually checked.

tobiasgrosser committed:
Merge remote-tracking branch 'origin/main' into llvm-riscv-merging-1

tobiasgrosser committed:
chore: rename DialectPrint to 'ToPrint'

tobiasgrosser committed:
chore: fix typo

tobiasgrosser commented:
I merge this now. We can iterate on the implementation in-tree.
